### ISSUE TITLE
Only first object saved in has_many with accepts_nested_attributes_for on both sides

### ISSUE THREAD

**Issue by tangopium** [rails/rails#26358]
Im working with Ruby 2.3.1 and Rails 4.2.7.1.

A have two objects which are connected through a has_many relation (1:n). In both objects I use `accepts_nested_attributes_for` for the relation, so the object look like this:

```
class Company < ActiveRecord::Base
   has_many :workers, inverse_of: :company
   accepts_nested_attributes_for :workers
end
```

```
class Worker < ActiveRecord::Base
   belongs_to :company, inverse_of: :workers
   accepts_nested_attributes_for :company
end
```

Lets say now I have 3 workers in the database that are not associated with another company (`company_id = nil`). If I initialize a new Company with `c = Company.new` and add the workers `c.workers << Worker.find(1)`, `c.workers << Worker.find(2)` and `c.workers << Worker.find(3)` and than save the company `c.save!`, only the `company_id` of the first worker in the has_many collection is updated.

I would expect that all three records get the `company_id` field set (and hence be associated to the company).

If I remove the `accepts_nested_attributes_for` from the worker, it works just fine, but of cause no attributes of the company will be saved, when saving a worker.

**Comment by vipulnsward** [comment]
Thanks for reporting, can try to recreate the scenario using [reproduction script](https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_gem.rb).

**Comment by tejasbubane** [comment]
Here is a reproduction script and I cannot reproduce the issue, all tests pass.

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  # Activate the gem you are reporting the issue against.
  gem "activerecord", "4.2.7.1"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :companies, force: true do |t|
  end

  create_table :workers, force: true do |t|
    t.integer :company_id
  end
end

class Company < ActiveRecord::Base
   has_many :workers, inverse_of: :company
   accepts_nested_attributes_for :workers
end

class Worker < ActiveRecord::Base
   belongs_to :company, inverse_of: :workers
   accepts_nested_attributes_for :company
end

class BugTest < Minitest::Test
  def setup
    worker_1 = Worker.create!
    worker_2 = Worker.create!
    worker_3 = Worker.create!

    @company = Company.create!
  end

  def test_association_stuff
    @company.workers << Worker.find(1)
    @company.workers << Worker.find(2)
    @company.workers << Worker.find(3)

    assert_equal 3, @company.workers.count
    assert_equal @company.id, Worker.find(1).company_id
    assert_equal @company.id, Worker.find(2).company_id
    assert_equal @company.id, Worker.find(3).company_id
  end
end

```

**Comment by tejasbubane** [comment]
Ok, this one reproduces the issue:

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  # Activate the gem you are reporting the issue against.
  gem "activerecord", "4.2.7.1"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :companies, force: true do |t|
  end

  create_table :workers, force: true do |t|
    t.integer :company_id
  end
end

class Company < ActiveRecord::Base
   has_many :workers, inverse_of: :company
   accepts_nested_attributes_for :workers
end

class Worker < ActiveRecord::Base
   belongs_to :company, inverse_of: :workers
   accepts_nested_attributes_for :company
end

class BugTest < Minitest::Test
  def setup
    worker_1 = Worker.create!
    worker_2 = Worker.create!
    worker_3 = Worker.create!

    @company = Company.new

    @company.workers << Worker.find(1)
    @company.workers << Worker.find(2)
    @company.workers << Worker.find(3)

    @company.save!
  end

  def test_three_workers
    assert_equal 3, @company.workers.count
  end

  def test_associated_with_first_worker
    assert_equal @company.id, Worker.find(1).company_id
  end

  def test_associated_with_second_worker
    assert_equal @company.id, Worker.find(2).company_id
  end

  def test_associated_with_third_worker
    assert_equal @company.id, Worker.find(3).company_id
  end
end
```

Result:

```
Finished in 0.142505s, 28.0692 runs/s, 28.0692 assertions/s.

  1) Failure:
BugTest#test_associated_with_third_worker [rails_26358.rb:74]:
Expected: 1
  Actual: nil


  2) Failure:
BugTest#test_associated_with_second_worker [rails_26358.rb:70]:
Expected: 3
  Actual: nil


  3) Failure:
BugTest#test_three_workers [rails_26358.rb:62]:
Expected: 3
  Actual: 1

4 runs, 4 assertions, 3 failures, 0 errors, 0 skips
```

**Comment by kamipo** [comment]
Fixed by #28155.
